vpr/src/device/rr_graph_obj.h

```diff
@@ -0,0 +1,647 @@
+    /* Only applicable to IPIN and OPIN, set the ptc_num for a node, which is the pin id in a logic block,
+     * See definition in t_type_descriptor data structure   
+     */
+    void set_node_pin_num(const RRNodeId& node, const short& pin_id);
+    /* Only applicable to CHANX and CHANY, set the ptc_num for a node, which is the track id in a routing channel,
+     * Routing channel is a group of routing tracks, each of which has a unique index
+     * An example 
+     *       Routing Channel 
+     *    +------------------+
+     *    |                  |
+     * ---|----------------->|----> track_id: 0
+     *    |                  |
+     *   ...   More tracks  ...
+     *    |                  |
+     * ---|----------------->|----> track_id: 1
+     *    |                  |
+     *    +------------------+
+     */
+    void set_node_track_num(const RRNodeId& node, const short& track_id);
+    /* Only applicable to SOURCE and SINK, set the ptc_num for a node, which is the class number in a logic block,
+     * See definition in t_type_descriptor data structure   
+     */
+    void set_node_class_num(const RRNodeId& node, const short& class_id);
+
+    /* Set the routing cost index for node, see node_cost_index() for details */
+    /* TODO the cost index should be changed to a StrongId!!! */
+    void set_node_cost_index(const RRNodeId& node, const short& cost_index);
+    /* Set the directionality for a node, only applicable to CHANX and CHANY */
+    void set_node_direction(const RRNodeId& node, const e_direction& direction);
+    /* Set the side for a node, only applicable to OPIN and IPIN */
+    void set_node_side(const RRNodeId& node, const e_side& side);
+    /* Set the RC information for a node */
+    void set_node_R(const RRNodeId& node, const float& R);
+    void set_node_C(const RRNodeId& node, const float& C);
+    /* Set the routing segment linked to a node, only applicable to CHANX and CHANY */
+    void set_node_segment(const RRNodeId& node, const RRSegmentId& segment_index);
+
+    /* Edge related */
+    /* classify the input edges of each node to be configurable (1st part) and non-configurable (2nd part) */
+    void partition_node_in_edges(const RRNodeId& node);
+    /* classify the output edges of each node to be configurable (1st part) and non-configurable (2nd part) */
+    void partition_node_out_edges(const RRNodeId& node);
+    void partition_in_edges();  /* classify the input edges of each node to be configurable (1st part) and non-configurable (2nd part) */
+    void partition_out_edges(); /* classify the output edges of each node to be configurable (1st part) and non-configurable (2nd part) */
+    void partition_edges();     /* classify the edges of each node to be configurable (1st part) and non-configurable (2nd part) */
+
+    /* Graph-level Clean-up, remove invalid nodes/edges etc. */
+    void compress();
+
+    /* Graph-level validation, ensure a valid RRGraph for routers */
+    bool validate();
```

Comment:
What is the difference between validate and check?

Reply:
Thanks for pointing this out! The validate() is the old method developed by Kevin during his first draft on RRGraph. The check() is the new method I have developed, in a very similar purpose. They are doing both fundamental checking. I have talked to Kevin and we think the two functions should be merged. 
Therefore, I rename the check() to be validate(), in order to differentiate from the external check_rr_graph(). The new validate() is based on the check() and includes the checking codes on vector sizes from old validate(). This makes validate() to be a more comprehensive checking function for data consistency.